---
title: "Identity Protocol for APCL project"
output: html_notebook
---

This notebook continues where filtering left off, moving the genepop into cervus and processing identity results.

## Import the genepop to cervus
The genepop should be stored either on google drive or github and downloaded onto the windows machine.

**Because I did not run the process genepop script first, I had to open the file and add a "pop" line between the contig header and the first sample info**

###Open Cervus and convert the genepop to ceruvs:
- Tools > Convert genotype file > Genepop to Cervus
- A window will open, navigate to the genepop file that has been downloaded to the windows machine
- Choose 2 digit format, do not use the first id as a population name
**(2893 individuals, 1447 loci)**

### Run an allele frequency analysis:
- Analysis > Allele frequency analysis
- make sure the converted file is in the genotype file field
* Yes Header Row
* Yes Read Locus Names
* ID in column 2
* First allele in column 3
* Type in number of loci from conversion (1447)
* Save as “….._AF"
* Leave all the output option boxes checked
 
### Run an identity analysis
* Genotype file should be autofilled with the convert file created above
* Yes Header Row
* Id in column 2
* First allele in column 3
* Do not test sexes separately
* Allele freq data should be autofilled with the AF file created above
* Save as “…._ID"
* Fill in minimum number of matching loci (80% of total = 1158)
* allow fuzzy matching with 10% mismatch (10% of total = 144)
* don’t show all comparisons - this generates a HUGE csv of every single pairwise comparison.

Upload the output to github

### Import the data into R

```{r}
library(tidyverse)
source("scripts/gen_helpers.R")

idcsv <- read_csv("data/seq03-33_identity/33-03_seq_identity_ID.csv", col_types = cols(
  `First ID` = col_character(),
  `Loci typed` = col_integer(),
  `Second ID` = col_character(),
  `Loci typed_1` = col_integer(),
  `Matching loci` = col_integer(),
  `Mismatching loci` = col_integer(),
  pID = col_double(),
  pIDsib = col_double(),
  Status = col_character()
))

# add mismatch rate
idcsv <- idcsv %>%
  mutate(mismatch_prop = `Mismatching loci`/`Matching loci`) %>% 
# order by number of matching loci
    arrange(`Matching loci`) %>% 
  # for the current run, there are extra .F on some names, remove them
  mutate(`First ID` = ifelse(grepl(".F", `First ID`), substr(`First ID`, 1, 10), `First ID`), 
         `Second ID` = ifelse(grepl(".F", `Second ID`), substr(`Second ID`, 1, 10), `Second ID`)) %>% 
  # remove APCL_ from IDs
  separate(`First ID`, c("spp", "first_id"), sep = "_") %>% 
  separate(`Second ID`, c("spp2", "second_id"), sep = "_") %>% 
  select(-spp, -spp2)


ggplot(idcsv, aes(x = `Matching loci`, y = mismatch_prop)) +
  geom_point() +
  theme_bw()

```

 ### Add metadata
```{r}
lab <- read_db("Laboratory")

samples <- lab %>%
  tbl("ligation") %>% 
  


# add Sample IDs
suppressWarnings(c1 <- labor %>% tbl("extraction") %>% select(extraction_id, sample_id))
suppressWarnings(c2 <- labor %>% tbl("digest") %>% select(digest_id, extraction_id))
c3 <- left_join(c2, c1, by = "extraction_id")
suppressWarnings(c4 <- labor %>% tbl("ligation") %>% select(ligation_id, digest_id))
c5 <- left_join(c4, c3, by = "digest_id") %>% collect()

### WAIT ###

# for First.ids 
lab1 <- c5
names(lab1) <- paste("First.", names(lab1), sep = "")

idcsv <- left_join(idcsv, lab1, by = c("First.ID" = "First.ligation_id"))

### WAIT ###

# For Second.IDs
lab2 <- c5
names(lab2) <- paste("Second.", names(lab2), sep = "")

idcsv <- left_join(idcsv, lab2, by = c("Second.ID" = "Second.ligation_id"))

```
 
