---
title: "Remove regenotyped samples"
output: html_notebook
---

This script is written to take the filtered genepop file from dDocent and
1) read the genepop file into R as a data frame
2) strip any named samples down to pure ligation number,
3) identify and remove re-genotyped samples based on number of loci (SNPs),
4) generate a new genepop file to be fed to cervus for identification of recaptures.

# Set up workspace ---------------------------------------------
```{r setup}
pacman::p_load(dplyr, stringr, readr, clownfish, here, install = FALSE)
# library(dplyr) 
# library(stringr)
# library(readr)
# library(clownfish)
# library(here)

# while db connection using helper file isn't working 
source("~/Documents/clownfish-pkg/R/db_connections.R")
leyte <- read_db("Leyte")
lab <- read_db("Laboratory")
```
# 1) Read the genepop  - double check genepop to make sure the word
# pop separates the header from the data on line 3 - no quotes
```{r}
# locate the genepop file and read as data frame
genfile <- "seq33_03_baits_only_SNPs.gen"
genedf <- read_genepop(here("filtering", genfile))%>% 
  rename(ligation_id = names) %>% 
  # 2) strip any named samples down to pure ligation number ---- 
  mutate(ligation_id = str_extract(ligation_id, "L\\d+"))
```
# 3) Remove samples with known issues -------------------------------------
# pull in the known issues table
```{r}
iss <- lab %>% tbl("known_issues") %>% collect()

# remove issues from largedf
genedf <- genedf %>%
  filter(!ligation_id %in% iss$ligation_id)
```


# Add sample IDs ----------------------------------------------------------

```{r}
samples <- samp_from_lig(genedf)

# Merge the two dataframes so that lig IDs match up -----------------------
largedf <- left_join(genedf, samples, by = "ligation_id") %>% 
  select(sample_id, ligation_id, everything()) # move the sample_id column to the beginning

# tests ------------------------------------------------------------------
# # make sure all of the Ligation ids have sample ids
# filter(largedf, is.na(sample_id))
# # returns 0 rows
# nrow(genedf) == nrow(largedf) # should be TRUE
# # # look for missing names
# setdiff(genedf$ligation_id, largedf$ligation_id) # should be character(0)

# should return integer(0)
# else trouble <- largedf %>% filter(is.na(largedf$sample_id)) - 
# should be on the known issues table

# TEST - make sure no more match the list
# j <- largedf %>%
#   filter(ligation_id %in% iss$ligation_id)
# nrow(j) # should return 0
# rm(j)
rm(genedf, genfile, samples, iss)
```



# Remove regenotyped samples ----------------------------------------------

# convert 0000 to NA in the genepop data
```{r}
largedf[largedf == "0000"] = NA
```


# # TEST - make sure there are no "0000" left
# which(largedf == "0000") # should return integer(0)

# count the number of loci per individual (have to use for loop)
```{r cache=TRUE}
for(i in 1:nrow(largedf)){
  largedf$numloci[i] <- sum(!is.na(largedf[i,]))
}

# # TEST - make sure all of the numloci were populated ----------------------
# which(is.na(largedf$numloci)) # should return integer(0)
```



```{r}

#run through all of the SampleIDs that are found more than once and keep the one with the most loci

# create a data frame of sample ids to keep or drop based on the number of loci present
regeno_drop <- largedf %>% 
  # select only the columns to identify which to keep
  select(sample_id, ligation_id, numloci) %>%
  group_by(sample_id) %>% 
  # create a maxloci column that holds the most loci each sample id carries
  mutate(maxloci = max(numloci), 
   # create a drop column that holds the decision to keep or drop that row
         drop = ifelse(numloci == maxloci, "KEEP", "DROP")) %>% 
  # keep only the columns that hold the maxloci and say "KEEP"
  filter(drop == "KEEP") %>% 
  ungroup() %>% 
  # keep only the columns needed to identify these "keeper rows"
  select(ligation_id)

# keep only the version of the sample id that has the most loci
noregeno <- largedf %>% 
  filter(ligation_id %in% regeno_drop$ligation_id)
```
# Some samples were not dropped because both regenotypes have the same number of loci.
```{r}
regenod <- noregeno %>%
  filter(duplicated(noregeno$sample_id)) %>%
  select(sample_id, ligation_id) %>% 
  distinct()

# drop the ligation_ids for these duplicated samples
noregeno <- noregeno %>% 
  filter(!ligation_id %in% regenod$ligation_id)




```





# convert all of the KEEPs to NAs 
```{r}
largedf$drop[largedf$drop == "KEEP"] <- NA
```


# create a new data frame with none of the "DROP" rows
```{r}
noregeno <- largedf[is.na(largedf$drop),]

# TEST - make sure no drop rows made it
which(noregeno$drop == "DROP") # should return integer(0)
# TEST - check to see if there are any regenos that were missed
noregeno$sample_id[duplicated(noregeno$sample_id)] # should return character(0)  
# If it doesn't, look deeper: noregeno[which(noregeno$SampleID == "APCL15_403"),], largedf[which(largedf$sample_ID == "APCL15_403"),]
```



```{r}
# remove the extra columns from noregeno
noregeno <- noregeno %>% 
  select(-numloci, -drop)

# convert all the NA genotypes to 0000
noregeno[is.na(noregeno)] = "0000"
# TEST - make sure there are no NA's left
which(is.na(noregeno)) # should return integer(0)

# TEST - compare the length of noregeno to the length of largedf
nrow(noregeno) == nrow(largedf) - k # 1569/1531 - should return TRUE


# # remove genotyped recaptures - only do this if you are ablsolutely sure you do not want to find new recapture events with this data
# leyte <- read_db("Leyte")
# recap <- leyte %>% tbl("clownfish") %>% 
#   filter(!is.na(cap_id)) %>% 
#   select(sample_id, cap_id) %>% 
#   collect()

# ########################################################################
# # # TEST - make sure a list was generated
# k <- nrow(recap)
# k # 277

```





<!-- noregeno$drop <- NA # place holder -->
#run through all of the SampleIDs that are found more than once and keep the one with the most loci
# for testing b <- 1
# for(i in 1:max(recap$cap_id)){
#   # recap_drop is the line number from noregeno that matches an ID in the regeno_match list
#   X <- recap$sample_id[recap$cap_id == recap$cap_id[i]]
#   recap_drop <- which(noregeno$sample_id %in% X)
#   # df is the data frame that holds all of the regenotyped versions of the sample, pulled from noregeno
#   df <- noregeno[recap_drop, ]  
#   # the row number of df with the largest number of loci (p-1 indicates the column)
#   keep <- which.max(df$numloci) 
#   # convert the df number to the row number of large df
#   c <- recap_drop[keep]
#   # convert the drop column of the row to keep to not na
#   df$drop[keep] <- "KEEP"
#   # convert the drop column of large df to not na
#   noregeno$drop[c] <- "KEEP"
#   
#   # find the row numbers of noregeno that need to be dropped
#   # test e <- 2
#   for(e in 1:nrow(df)){
#     if(is.na(df$drop[e])){
#       f <-recap_drop[e]
#       noregeno$drop[f] <- "DROP"
#     }
#   }
# }

# convert all of the KEEPs to NAs 
for(g in 1:nrow(noregeno)){
  if(!is.na(noregeno$drop[g]) && noregeno$drop[g]=="KEEP"){
    noregeno$drop[g] <- NA
  }
}

<!-- # create a new data frame with none of the "DROP" rows -->
<!-- noregeno <- noregeno[is.na(noregeno$drop),] -->
<!-- # TEST - make sure no drop rows made it -->
<!-- which(noregeno$drop == "DROP") # should return integer(0) -->
<!-- # TEST - check to see if there are any regenos that were missed -->
<!-- noregeno_match <- noregeno$sample_id[duplicated(noregeno$sample_id)] -->
<!-- noregeno_match # should return character(0)   -->
<!-- # If it doesn't, look deeper: noregeno[which(noregeno$SampleID == "APCL15_403"),], largedf[which(largedf$sample_ID == "APCL15_403"),] -->

<!-- # remove the extra columns from noregeno -->
<!-- noregeno [,c("extraction_ID")] <- NULL -->
<!-- noregeno [,c("digest_ID")] <- NULL -->
<!-- noregeno [,c("numloci")] <- NULL -->
<!-- noregeno [,c("drop")] <- NULL -->

<!-- # convert all the NA genotypes to 0000 -->
<!-- noregeno[is.na(noregeno)] = "0000" -->
<!-- # TEST - make sure there are no NA's left -->
<!-- which(is.na(noregeno)) # should return integer(0) -->

<!-- # TEST - compare the length of noregeno to the length of largedf -->
<!-- nrow(noregeno) == nrow(largedf) - k # 1569/1531 - should return TRUE -->



# 4) Output genepop file --------------------------------------------------

```{r}
# Build the genepop components
msg <- c("This genepop file was generated using a script called process_genepop.Rmd written by Michelle Stuart ")
write_lines(msg, path = str_c(here("data", "seq33-03_noregeno.gen"), sep = ""), append = F)

# find all of the column names that contain contig and collapse them into a comma separated string
loci <-  str_c(names(select(noregeno, contains("contig"))), collapse = ",")
write_lines(loci, path = str_c(here("data", "seq33-03_noregeno.gen"), sep = ""), append = T)

pop <- "pop"
write_lines(pop, path = str_c(here("data", "seq33-03_noregeno.gen"), sep = ""), append = T)

gene <- vector()
sample <- vector()
for (i in 1:nrow(noregeno)){
  gene[i] <- str_c(select(noregeno[i,], contains("contig")), collapse = " ")
  sample[i] <- str_c(noregeno$names[i], gene[i], sep = ", ")
  write_lines(sample[i], path = str_c(here("data", "seq33-03_noregeno.gen"), sep = ""), append = T)
}
```



